src/neurostatslib/utils.py

```diff
@@ -4,44 +4,263 @@
+    time_series: Union[Iterable[NDArray], NDArray, Iterable[jnp.ndarray], jnp.ndarray],
+) -> List[jnp.ndarray]:
+    """
+    Convolve trial time series with a basis matrix.
+
+    This function checks if all trials have the same duration. If they do, it uses a fast method
+    to convolve all trials with the basis matrix at once. If they do not, it falls back to convolving
+    each trial individually.
+
+    Parameters
+    ----------
+    basis_matrix :
+        The basis matrix with which to convolve the trials. Shape (window_size, n_basis_funcs).
+    time_series :
+        The time series of trials to convolve with the basis matrix. It should be a list of 2D arrays,
+        where each array represents a trial and its second dimension matches the first dimension
+        of the basis_matrix. Each trial has shape (n_timebins_trial, n_neurons).
+
+    Returns
+    -------
+    :
+        The convolved trials. It is a list of 3D arrays, where each array represents a convolved trial.
+        Each element of the list will have shape (n_timebins_trial - window_size - 1, n_neurons, n_basis_funcs).
+
+    Raises
+    ------
+    ValueError
+        - If basis_matrix is not a 2D array-like object.
+        - If trials_time_series is not an iterable of 2D array-like objects.
+        - If trials_time_series contains empty trials.
+        - If the number of time points in each trial is less than the window size.
+    """
+
+    basis_matrix = jnp.asarray(basis_matrix)
+    # check input size
+    if basis_matrix.ndim != 2:
+        raise ValueError("basis_matrix must be a 2 dimensional array-like object.")
+    elif 0 in basis_matrix.shape:
+        raise ValueError(
+            "Empty basis_matrix provided. "
+            f"The shape of basis_matrix is {basis_matrix.shape}!"
+        )
+    try:
+        # this should fail for variable trial length
+        time_series = jnp.asarray(time_series)
+        if time_series.ndim != 3:
+            raise ValueError
+
+    except ValueError:
+        # convert each trial to array
+        time_series = [jnp.asarray(trial) for trial in time_series]
+        if not check_dimensionality(time_series, 2):
+            raise ValueError(
+                "time_series must be an iterable of 2 dimensional array-like objects."
+            )
+
+    if any(k == 0 for trial in time_series for k in trial.shape) or (
+        len(time_series) == 0
+    ):
+        raise ValueError("time_series should not contain empty trials!")
+
+    # Check window size
+    ws = len(basis_matrix)
+    if any(trial.shape[0] < ws for trial in time_series):
+        raise ValueError(
+            "Insufficient trial duration. The number of time points in each trial must "
+            "be greater or equal to the window size."
+        )
+
+    # Check if all trials have the same duration # check if needed
+    same_dur = time_series.ndim == 3 if isinstance(time_series, jnp.ndarray) else False
+
+    if same_dur:
```

Comment:
```suggestion
    if isinstance(time_series, jnp.ndarray) and time_series.ndim == 3 :
```
this feels cleaner to me, but your original way is fine too, just a little confusing to understand

Reply:
I checked and tested the logic, if the conversion to array worked, I have already checked that time_series has 3 dimensions. I simplified the if close to just isinstance